### PR TITLE
WIP: Abstracts

### PR DESCRIPTION
**Pull request by nop33**

### CODE REVIEW

File: indico/modules/events/abstracts/models/notification_templates.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def _get_next_position(context):
+    """Get the next abstracts template position for the event."""
+    event_id = context.current_parameters['event_id']
+    res = db.session.query(db.func.max(AbstractsNotificationTemplate.position)).filter_by(event_id=event_id).one()
+    return (res[0] or 0) + 1
+
+
+class AbstractsNotificationTemplate(NotificationTemplateBase, db.Model):
```
Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: Why not inherit the base class from `db.Model` (add `__abstract__ = True` to avoid the base from creating its own table)?

File: indico/modules/events/abstracts/models/notification_templates.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    """Represents an email notification template for abstracts"""
+
+    __tablename__ = 'notification_templates'
+    __auto_table_args = ({'schema': 'event_abstracts'})
```
Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: those parentheses are are useless/misleading here - you only need a tuple if you want both a dict and non-dict tableargs

File: indico/modules/events/abstracts/models/notification_templates.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    #: Whether to include co-authors' email addresses as CC for the
+    #: notification emails
+    include_coauthors = db.Column(
+        db.Boolean,
+        nullable=False,
+        default=False
+    )
+    #: The relative position of the template in the list of abstract
+    #: notification templates
+    position = db.Column(
```
Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: This makes more sense in the base class. See if you can get the class from `context` in the default value callback

File: indico/modules/events/abstracts/models/notification_templates.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        db.Integer,
+        nullable=False,
+        default=_get_next_position
+    )
+
+    @property
+    def to_emails(self):
+        emails = []
```
Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: should be a set

File: indico/modules/events/abstracts/models/notification_templates.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        if self.include_authors:
+            # TODO: get list of abstract's authors
+            authors_emails = []
+            emails.extend(authors_emails)
+            pass
```
Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: :monkey:

File: indico/modules/events/abstracts/models/notification_templates.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        if self.include_submitters:
+            # TODO: get list of abstract's submitters
```
Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: fyi, this is not a list of submitters - it's the one person who submitted the abstract

File: indico/modules/events/abstracts/models/notification_templates.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        return emails
+
+    @property
+    def cc_emails(self):
+        emails = []
+        if self.include_coauthors:
+            # TODO: get list of abstract's co-authors
+            coauthors_emails = []
+            emails.extend(coauthors_emails)
+        if self.extra_cc_emails:
```
Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: thie belongs in the base (initialize emails with super at the beginning of the method here)

File: indico/modules/events/abstracts/models/notification_templates.py (hunk 2 above)
Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: use regular `__table_args__` if you don't need to get some from a mixin/base. right now they don't do anything because you don't have the regular one with the `auto_table_args` call.

File: indico/util/notification_templates.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+from sqlalchemy.ext.declarative import declared_attr
+
+from indico.core.db import db
+
+
+class NotificationTemplateBase(object):
+    """Base class for notification templates."""
+
+    #: The ID of the template
+    @declared_attr
```
Comment (ThiefMaster) [inline indico/util/notification_templates.py]: you can use regular column definitions - no need for `declared_attr` here

File: indico/util/notification_templates.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    def id(cls):
+        return db.Column(
+            db.Integer,
+            primary_key=True
+        )
+
+    #: The ID of the event
+    @declared_attr
+    def event_id(cls):
```
Comment (ThiefMaster) [inline indico/util/notification_templates.py]: there should be an `event_new` relationship. for the backref use a class attribute which you set in the subclass

File: indico/util/notification_templates.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+    #: List of email addresses to send the notification email
+    @property
+    def to_emails(self):
+        raise NotImplementedError
+
+    #: List of email addresses to be added as CC in the notification email
+    @property
+    def cc_emails(self):
+        raise NotImplementedError
```
Comment (ThiefMaster) [inline indico/util/notification_templates.py]: this *can* be implemented here

File: indico/util/notification_templates.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            nullable=True,
+            default=[],
+        )
+
+    #: List of email addresses to send the notification email
+    @property
+    def to_emails(self):
+        raise NotImplementedError
+
+    #: List of email addresses to be added as CC in the notification email
```
Comment (ThiefMaster) [inline indico/util/notification_templates.py]: use a proper docstring - only things that are neither functions nor classes need the `#:` syntax

File: migrations/versions/201608011329_4bf6f41baafa_add_notification_templates_models.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        sa.Column('event_id', sa.Integer(), nullable=False),
+        sa.Column('reply_to_address', sa.String(), nullable=False),
+        sa.Column('extra_cc_emails', postgresql.ARRAY(sa.String()), nullable=True),
+        sa.Column('stop_on_match', sa.Boolean(), nullable=False),
+        sa.Column('conditions', postgresql.JSON(), nullable=False),
+        sa.Column('id', sa.Integer(), nullable=False),
+        sa.Column('subject', sa.String(), nullable=False),
+        sa.ForeignKeyConstraint(['event_id'], [u'events.events.id'],
+                                name=op.f('fk_notification_templates_event_id_events')),
+        sa.PrimaryKeyConstraint('id', name=op.f('pk_notification_templates'))
```
Comment (ThiefMaster) [inline migrations/versions/201608011329_4bf6f41baafa_add_notification_templates_models.py]: no name needed either

File: migrations/versions/201608011329_4bf6f41baafa_add_notification_templates_models.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        sa.Column('title', sa.String(), nullable=False),
+        sa.Column('event_id', sa.Integer(), nullable=False),
+        sa.Column('reply_to_address', sa.String(), nullable=False),
+        sa.Column('extra_cc_emails', postgresql.ARRAY(sa.String()), nullable=True),
+        sa.Column('stop_on_match', sa.Boolean(), nullable=False),
+        sa.Column('conditions', postgresql.JSON(), nullable=False),
+        sa.Column('id', sa.Integer(), nullable=False),
+        sa.Column('subject', sa.String(), nullable=False),
+        sa.ForeignKeyConstraint(['event_id'], [u'events.events.id'],
+                                name=op.f('fk_notification_templates_event_id_events')),
```
Comment (ThiefMaster) [inline migrations/versions/201608011329_4bf6f41baafa_add_notification_templates_models.py]: remove the name kwarg and the `u` prefix

File: migrations/versions/201608011329_4bf6f41baafa_add_notification_templates_models.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        sa.PrimaryKeyConstraint('id', name=op.f('pk_notification_templates'))
+    )
+    op.create_index(op.f('ix_notification_templates_event_id'), 'notification_templates', ['event_id'], unique=False)
```
Comment (ThiefMaster) [inline migrations/versions/201608011329_4bf6f41baafa_add_notification_templates_models.py]: `index=True` in the column definition instead

File: migrations/versions/201608011329_4bf6f41baafa_add_notification_templates_models.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def downgrade():
+    op.drop_index(op.f('ix_notification_templates_event_id'), table_name='notification_templates')
```
Comment (ThiefMaster) [inline migrations/versions/201608011329_4bf6f41baafa_add_notification_templates_models.py]: not needed, it's deleted with the table

File: migrations/versions/201608011329_4bf6f41baafa_add_notification_templates_models.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    op.drop_table('notification_templates')
```
Comment (ThiefMaster) [inline migrations/versions/201608011329_4bf6f41baafa_add_notification_templates_models.py]: `schema='event_abstracts'` missing here and during event creation (it'd have been there had the tableargs been set correctly ;))

File: indico/util/notification_templates.py (hunk 8 above)
Comment (nop33) [inline indico/util/notification_templates.py]: I had regular column definitions, but when I created the revision it game the error `sqlalchemy.exc.InvalidRequestError: Columns with foreign keys to other columns must be declared as @declared_attr callables on declarative mixin classes.`

Comment (ThiefMaster) [inline indico/util/notification_templates.py]: ah right, forgot about the FK columns. in that case it makes sense to use it for all of them to have them in a less strange order

Comment (nop33) [inline indico/util/notification_templates.py]: All of them? You mean also the ones in the `AbstractNotificationTemplate`?

File: indico/util/notification_templates.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+    #: The ID of the template
+    @declared_attr
+    def id(cls):
+        return db.Column(
+            db.Integer,
+            primary_key=True
+        )
+
+    #: The ID of the event
```
Comment (ThiefMaster) [inline indico/util/notification_templates.py]: `"""..."""` inside the method. and the docstring is wrong ;)

I wouldn't put docstrings on columns whose meaning is obvious. I know we did it in some models but then stopped doing it in most models afterwards. It's obvious what the `id` column does after all

File: indico/util/notification_templates.py (hunk 8 above)
Comment (ThiefMaster) [inline indico/util/notification_templates.py]: No, just here. Basically what you did.

Comment (ThiefMaster) [inline indico/util/notification_templates.py]: Actually, *not* using `declared_attr` for the most basic ones might give you a nicer column order (with id being the first one in the table). No technical difference, just looks nicer in pgadmin etc. If you change it make sure to regenerate the alembic revision to see whether the order actually changed

File: indico/modules/events/abstracts/models/notification_templates.py (hunk 6 above)
Comment (nop33) [inline indico/modules/events/abstracts/models/notification_templates.py]: The submitters can be more than one

Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: How? We had a look at the code on Friday and saw that this is not the people who can submit stuff. It's the person who submitted the abstract..

Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: https://github.com/indico/indico/blob/master/indico/MaKaC/review.py#L1329
https://github.com/indico/indico/blob/master/indico/MaKaC/review.py#L3258-L3262

Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: So it's just awful naming in the legacy code (and should actually use singular `submitter` in any new code)

Comment (nop33) [inline indico/modules/events/abstracts/models/notification_templates.py]: ah, okay, even in the UI we say "Submitters" next to the checkbox. I'll update.

Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: Yeah, exactly what I meant with awful naming ;)

File: indico/util/notification_templates.py (hunk 8 above)
Comment (nop33) [inline indico/util/notification_templates.py]: By "most basic ones", you mean what? The `String`, `Text`, `Boolean`, ... columns?

File: indico/util/notification_templates.py (hunk 9 above)
Comment (ThiefMaster) [inline indico/util/notification_templates.py]: mainly id and title

File: indico/util/notification_templates.py (hunk 8 above)
Comment (ThiefMaster) [inline indico/util/notification_templates.py]: mainly id and title

Comment (nop33) [inline indico/util/notification_templates.py]: But we just said that we **should** use `declared_attr` for `id`

Comment (ThiefMaster) [inline indico/util/notification_templates.py]: event_id needs it because of the FK

basically sqlalchemy takes regular columns first and then ones from declared_attr. it looks a bit weird when you have the PK ID not as the first column IMO which is why i would not use declared_attr for this one

File: indico/util/notification_templates.py (hunk 11 above)
Comment (nop33) [inline indico/util/notification_templates.py]: So, `"""..."""` instead of `:# ...`?
Then we should change here at some point too :stuck_out_tongue: 
https://github.com/indico/indico/blob/master/indico/core/storage/models.py#L62

Comment (ThiefMaster) [inline indico/util/notification_templates.py]: yep ;)

File: indico/util/notification_templates.py (hunk 8 above)
Comment (nop33) [inline indico/util/notification_templates.py]: I see. But in this case, only the `title` will be without `declared_attr`, since the `id` and `event_id` need it.

File: indico/modules/events/abstracts/models/notification_templates.py (hunk 3 above)
Comment (nop33) [inline indico/modules/events/abstracts/models/notification_templates.py]: So, the definition of `_get_next_position` stays in the abstracts module and the `position` will be part of `NotificationTemplateBase`. And then I can probably do `cls._get_next_position` in the default value callback, no?

Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: Yes - but if you can just get the model from the context it'd be nicer IMO

Comment (nop33) [inline indico/modules/events/abstracts/models/notification_templates.py]: Hm, trying to figure out how to get the model from the context. In models all uses I can find of `context` is mainly to get the `event_id` from the `current_parameters`.

Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: http://docs.sqlalchemy.org/en/latest/core/defaults.html#context-sensitive-default-functions

You'll have to break there with a debugger and check what else the context object contains. It doesn't seem to be documented

Comment (nop33) [inline indico/modules/events/abstracts/models/notification_templates.py]: I can find the correct table name in the context, but no info about the actual `AbstractsNotificationTemplate` model...

Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: If you have the table name you can look up the model like this (or ask in #sqlalchemy if there's a better way ;)):

```
>>> next((x for x in db.Model._decl_class_registry.itervalues() if hasattr(x, '__table__') and x.__table__.fullname == 'events.events'), None)
indico.modules.events.models.events.Event
```

File: indico/modules/events/abstracts/models/notification_templates.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Indico; if not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import unicode_literals
+
+from indico.core.db import db
+from indico.util.notification_templates import NotificationTemplateBase
+
+
+class AbstractsNotificationTemplate(NotificationTemplateBase):
```
Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: AbstractNotificationTemplate

File: indico/modules/events/notification_templates/models/templates.py [hunk 19]
```diff
@@ -0,0 +1,137 @@
+# This file is part of Indico.
```
Comment (ThiefMaster) [inline indico/modules/events/notification_templates/models/templates.py]: Does it really make sense to add a whole module just for the base model? I'd rather use `indico/modules/events/notification_templates.py` - it's not really a module if it has nothing besides some model.

File: indico/modules/events/abstracts/models/notification_templates.py [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Indico; if not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import unicode_literals
+
+from indico.core.db import db
+from indico.modules.events.notification_templates.models.templates import NotificationTemplateBase
```
Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: Does it really make sense to add a whole module just for the base model? I'd rather use `indico/modules/events/notification_templates.py` - it's not really a module if it has nothing besides some model.

File: indico/modules/events/abstracts/controllers/management.py [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+from MaKaC.webinterface.rh.base import RH
+from MaKaC.webinterface.rh.conferenceModif import RHConferenceModifBase
+
+
+class RHManageAbstractsBase(RHConferenceModifBase):
+    """Base class for all abstracts management RHs"""
+    CSRF_ENABLED = True
+
+    def _checkParams(self, params):
+        RHConferenceModifBase._checkParams(self, params)
```
Comment (ThiefMaster) [inline indico/modules/events/abstracts/controllers/management.py]: no need to override this if you just call the parent method

File: indico/modules/events/timetable/forms.py [hunk 22]
```diff
@@ -57,7 +56,10 @@ def __init__(self, *args, **kwargs):
             kwargs.setdefault('time', self._get_default_time())
             defaults = kwargs.get('obj') or FormDefaults()
             if 'duration' not in defaults:
-                defaults.duration = self._default_duration
+                if self._entry_type == TimetableEntryType.CONTRIBUTION and self.session_block:
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/forms.py]: This doesn't seem to belong in this PR. Rebase issue?

File: indico/web/flask/blueprints/event/management/abstracts.py [hunk 23]
```diff
@@ -19,144 +19,144 @@
 
 
 # Setup
-event_mgmt.add_url_rule('/call-for-abstracts/setup/', 'confModifCFA', conferenceModif.RHConfModifCFA)
-event_mgmt.add_url_rule('/call-for-abstracts/setup/toggle', 'confModifCFA-changeStatus',
+event_mgmt.add_url_rule('/call-for-abstracts-old/setup/', 'confModifCFA', conferenceModif.RHConfModifCFA)
```
Comment (ThiefMaster) [inline indico/web/flask/blueprints/event/management/abstracts.py]: no need to add `-old` here - the new CFA urls use `/abstracts/` instead of `/call-for-abstracts/`

File: migrations/versions/201608011709_4d4b95748173_add_notification_templates_models.py [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+        sa.Column('id', sa.Integer(), nullable=False),
+        sa.Column('title', sa.String(), nullable=False),
+        sa.Column('include_authors', sa.Boolean(), nullable=False),
+        sa.Column('include_submitter', sa.Boolean(), nullable=False),
+        sa.Column('include_coauthors', sa.Boolean(), nullable=False),
+        sa.Column('body', sa.Text(), nullable=False),
+        sa.Column('reply_to_address', sa.String(), nullable=False),
+        sa.Column('subject', sa.String(), nullable=False),
+        sa.Column('event_id', sa.Integer(), nullable=False, index=True),
+        sa.Column('extra_cc_emails', postgresql.ARRAY(sa.String()), nullable=True),
```
Comment (ThiefMaster) [inline migrations/versions/201608011709_4d4b95748173_add_notification_templates_models.py]: this shouldn't be nullable

File: indico/modules/events/notification_templates.py [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+            JSON,
+            nullable=False
+        )
+
+    @declared_attr
+    def extra_cc_emails(cls):
+        """List of extra email addresses to be added as CC in the notification email"""
+        return db.Column(
+            ARRAY(db.String),
+            nullable=True,
```
Comment (ThiefMaster) [inline indico/modules/events/notification_templates.py]: `False`

File: indico/modules/events/abstracts/controllers/management.py (hunk 21 above)
Comment (nop33) [inline indico/modules/events/abstracts/controllers/management.py]: I know, but I it's soon gonna change

File: migrations/versions/201608011709_4d4b95748173_add_notification_templates_models.py (hunk 24 above)
Comment (nop33) [inline migrations/versions/201608011709_4d4b95748173_add_notification_templates_models.py]: Why not? Isn't it **extra** addresses?

Comment (ThiefMaster) [inline migrations/versions/201608011709_4d4b95748173_add_notification_templates_models.py]: Yes, and if you don't have any you have an empty array, not NULL

Comment (nop33) [inline migrations/versions/201608011709_4d4b95748173_add_notification_templates_models.py]: True, didn't think of that

File: indico/core/db/sqlalchemy/util/models.py [hunk 26]
```diff
@@ -419,3 +419,9 @@ def _expr(cls):
         return getattr(cls, own_attr_name)
 
     return hybrid_property(_get, _set, expr=_expr)
+
+
+def get_model_by_table_name(table_name):
```
Comment (ThiefMaster) [inline indico/core/db/sqlalchemy/util/models.py]: `"""Get the model class based on the name of the table"""`

File: indico/modules/events/abstracts/controllers/management.py (hunk 21 above)
Comment (ThiefMaster) [inline indico/modules/events/abstracts/controllers/management.py]: In this one? I don't think so... this is the base class for all management RHs related to abstracts.

File: indico/modules/events/abstracts/models/notification_templates.py [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+    events_backref_name = 'abstract_notification_templates'
+
+    #: Whether to include authors' email addresses as To for the notification
+    #: emails
+    include_authors = db.Column(
+        db.Boolean,
+        nullable=False,
+        default=False
+    )
+    #: Whether to include submitter's email address as To for the
```
Comment (ThiefMaster) [inline indico/modules/events/abstracts/models/notification_templates.py]: *the* submitter's email address

File: indico/modules/events/registration/util.py [hunk 28]
```diff
@@ -422,3 +425,185 @@ def build_registration_api_data(registration):
...
+            })
+        ])
+        self.basic_items = ('title', 'first_name', 'last_name', 'email', 'position', 'affiliation', 'address', 'phone',
+                            'country')
+        self.special_items = tuple(self.special_items_info)
+        self.list_config = self._get_config()
+
+    def _get_non_custom_columns(self, ids):
+        """
+        Retrieve information needed for the header of the non custom columns (basic and special).
```
Comment (ThiefMaster) [inline indico/modules/events/registration/util.py]: wrap after "custom" - that line is way too long for text (comment/docstring)

File: indico/modules/events/registration/util.py [hunk 29]
```diff
@@ -422,3 +425,185 @@ def build_registration_api_data(registration):
...
+
+        Returns a list of ``{'id': ..., 'caption': ...}`` dictionaries.
```
Comment (ThiefMaster) [inline indico/modules/events/registration/util.py]: `:return: a list of  ``{'id': ..., 'caption': ...}`` dicts`

File: indico/modules/events/registration/util.py [hunk 30]
```diff
@@ -422,3 +425,185 @@ def build_registration_api_data(registration):
...
+                'title': _('Check-in date'),
+                'id': 'checked_in_date'
+            })
+        ])
+        self.basic_items = ('title', 'first_name', 'last_name', 'email', 'position', 'affiliation', 'address', 'phone',
+                            'country')
+        self.special_items = tuple(self.special_items_info)
+        self.list_config = self._get_config()
+
+    def _get_non_custom_columns(self, ids):
```
Comment (ThiefMaster) [inline indico/modules/events/registration/util.py]: "non custom columns"? that's an odd name

Comment (nop33) [inline indico/modules/events/registration/util.py]: yeah well... It's a very complicated situation.. We have the `custom` columns which come from the custom registration fields, the `special` columns which are info like checkin, registration state, date, etc., and `basic` which are personal info about the registrant. The above function returns the columns that are not custom so basically the `special` and `basic`

Comment (ThiefMaster) [inline indico/modules/events/registration/util.py]: "static" columns maybe?

Comment (nop33) [inline indico/modules/events/registration/util.py]: Well, they are not really static either, since they can be visible or hidden exactly like the custom columns.

Comment (ThiefMaster) [inline indico/modules/events/registration/util.py]: static as in not dynamic (fields that are created by the user)

Comment (nop33) [inline indico/modules/events/registration/util.py]: Hm, okay, I see. I will put a docstring as well.

Comment (nop33) [inline indico/modules/events/registration/util.py]: nvm, there is a docstring already :tongue:
